include/fiction/algorithms/simulation/sidb/assess_physical_population_stability.hpp

```diff
@@ -0,0 +1,358 @@
+#define FICTION_ASSESS_PHYSICAL_POPULATION_STABILITY_HPP
+
+#include "fiction/algorithms/simulation/sidb/convert_potential_to_distance.hpp"
+#include "fiction/algorithms/simulation/sidb/quickexact.hpp"
+#include "fiction/algorithms/simulation/sidb/sidb_simulation_parameters.hpp"
+#include "fiction/layouts/cell_level_layout.hpp"
+#include "fiction/traits.hpp"
+#include "fiction/types.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <iterator>
+#include <limits>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+namespace fiction
+{
+
+/**
+ * Possible types of charge transitions that can occur in an SiDB layout. These transitions represent
+ * changes in the charge state of SiDBs, including transitions from neutral to negative, negative to neutral, neutral to
+ * positive, and positive to neutral.
+ */
+enum class transition_type
+{
+    /**
+     * SiDB is neutrally charged, but is closest to being negatively charged.
+     */
+    NEUTRAL_TO_NEGATIVE,
+    /**
+     * SiDB is negatively charged, but is closest to being neutrally charged.
+     */
+    NEGATIVE_TO_NEUTRAL,
+    /**
+     * SiDB is neutrally charged, but is closest to being positively charged.
+     */
+    NEUTRAL_TO_POSITIVE,
+    /**
+     * SiDB is positively charged, but is closest to being neutrally charged.
+     */
+    POSITIVE_TO_NEUTRAL,
+};
+
+template <typename Lyt>
+struct population_stability_information
+{
+    /**
+     * SiDB cell which is closest to a charge transition.
+     */
+    typename Lyt::cell critical_cell{};
+    /**
+     * Charge transition from the current charge state to the closest one.
+     */
+    transition_type transition_from_to{};
+    /**
+     * Absolute electrostatic potential (unit: V) required for the charge state transition.
+     */
+    double minimum_potential_difference_to_transition{};
+    /**
+     * Distance (unit: nm) corresponding to the minimum potential difference.
+     */
+    double distance_corresponding_to_potential{};
+    /**
+     * Total electrostatic energy (unit: eV) of given charge distribution.
+     */
+    double system_energy{};
+};
+
+/**
+ * This struct stores the parameters required to assess the population stability.
+ */
+struct assess_physical_population_stability_params
+{
+    /**
+     * Parameters for the electrostatic potential.
+     */
+    sidb_simulation_parameters physical_parameters{};
+    /**
+     * The precision level for the conversion from the minimum potential difference to the corresponding
+     * distance.
+     */
+    uint64_t precision_for_distance_corresponding_to_potential = 2;
+};
+
+/**
+ * This struct represents the electrostatic energy and charge index of a charge distribution.
+ */
+struct energy_and_charge_index
+{
+    /**
+     * Electrostatic energy of the charge distribution.
+     */
+    double energy;
+    /**
+     * Charge index of the charge distribution.
+     */
+    uint64_t charge_index;
+};
```

Comment:
If I'm not mistaken, this `struct` is purely used in the internal context of `assess_physical_population_stability_impl`. Why not make it a `private` member of `assess_physical_population_stability_impl`?